// amity/lib/feathers-mongo-databases.js
var async = require('async');

/**
 * A feathers service to manage CRUD on MongoDB databases.
 * Pass in a connected MongoClient and it will allow you to create,
 * read, update, and delete databases on the client.
 *
 * @param  {[type]} db [description]
 * @return {[type]}    [description]
 */
module.exports = function(db) {

	var adminDB = db.admin();

	var dbService = {
		find: function(params, callback) {
			// Get a list of all databases on the server.
		  adminDB.listDatabases(function(err, dbs){

		  	// Async function to add stats to each db.
		  	var addStats = function(database, callback){
		  		// Switch databases
		  		var changeDB = db.db(database.name);
		  		// Get the stats
		  		changeDB.stats(function(err, stats){
		  			// Remove 1 from collection count. Not sure why it's 1 off.
		  			stats.collections--;
		  			// Add the stats to the corresponding database.
		  			database.stats = stats;
			  		callback();
		  		});
		  	}
		  	async.each(dbs.databases, addStats, function(err){
	  	    if (err) {
	  	    	callback(err);
	  	    } else {
					  callback(null, dbs.databases);
	  	    }
		  	});
		  });
		},

		create: function(data, params, callback) {
			console.log(data);
			callback(null, data);
		},

		update: function(id, data, params, callback) {
			console.log(data);
			callback(err, data);
		},

		remove: function(id, params, callback) {
			console.log(data);
			callback(err, data);
		},

		setup: function(app) {
			console.log(path);
			// this.service = app.service.bind(app);
		}
	};

	return dbService;
};
